core/src/main/scala/loadbalancers/Aws.scala

```diff
@@ -68,40 +65,179 @@ final class Aws(cfg: Infrastructure.Aws) extends (LoadbalancerOp ~> IO) {
 
   def delete(lb: Datacenter.LoadbalancerDeployment): IO[Unit] = {
     val name = loadbalancerName(lb.loadbalancer.name, lb.loadbalancer.version, lb.hash)
-    deleteELB(name) *> deleteASG(name)
+    for {
+      lbarn <- getLoadBalancerArn(name)
+      _ <- deleteNLB(lbarn)
+
+      targetGroups <- getTargetGroupsForNLB(lbarn)
+      _ <- deleteTargetGroups(targetGroups)
+
+      _ <- deleteASG(name)
+    } yield ()
   }
 
   def launch(lb: Manifest.Loadbalancer, v: MajorVersion, ns: NamespaceName, p: Plan, hash: String): IO[DNSName] = {
     val name = loadbalancerName(lb.name, v, hash)
-    log.debug(s"caling aws client to launch $name")
+    log.debug(s"calling aws client to launch $name")
     val size = asgSize(p)
-    createELB(name, lb.routes.map(_.port), cfg.lbScheme) <* createASG(name, ns, size)
+
+
+    for {
+      loadbalancer <- createNLB(name, cfg.lbScheme)
+
+      targetGroups <- createTargetGroups(name, lb.routes.map(_.port))
+      _ <- createListeners(loadbalancer, targetGroups)
+
+      _ <- createASG(name, ns, size)
+      _ <- attachASGToTargetGroups(name, targetGroups)
+    } yield loadbalancer.getDNSName
   }
 
+  // The target group has this restriction where
+  // the name of the target group can't have "double dashes". I want the loadbalancer name
+  // to match the target group name, so the loadbalancer won't have the double dashes either.
   def loadbalancerName(name: String, v: MajorVersion, hash: String) =
-    s"$name--${v.minVersion.toExternalString}--${hash}"
+    s"$name-${v.minVersion.toExternalString}-${hash}"
+
+  def getLoadBalancerArn(name: String): IO[String] = {
```

Comment:
Perhaps:

```
def getLoadBalancerArn(name: String): IO[String] = {
  val describeLoadBalancerRequest = new DescribeLoadBalancersRequest().withNames(name)
  (for {
    a <- IO(cfg.nlb.describeLoadBalancers(describeLoadBalancerRequest))
    b  = a.getLoadBalancers.asScala.toList
    c  = loadbalancerList.head.getLoadBalancerArn
  } yield ()).recoverWith{
    case e: Exception =>
      log.error(s"aws call to get loadbalancer ARN ${describeLoadBalancerRequest.getNames.asScala.head} failed.")
      IO.raiseError(FailedLoadbalancerDeploy(s"could not find loadbalancer and therefore could not delete loadbalancer", e.getMessage))
  }
}
```

I haven't tried this, but something along those lines to avoid the map map map situation

Reply:
I've used that map/map pattern a few places. This looks way nice :smile: